src/librustc_ast/ast.rs

```diff
@@ -309,19 +310,48 @@ pub type GenericBounds = Vec<GenericBound>;
 /// Specifies the enforced ordering for generic parameters. In the future,
 /// if we wanted to relax this order, we could override `PartialEq` and
 /// `PartialOrd`, to allow the kinds to be unordered.
-#[derive(PartialEq, Eq, PartialOrd, Ord, Hash, Clone, Copy)]
+#[derive(PartialEq, Eq, Hash, Clone, Copy)]
 pub enum ParamKindOrd {
     Lifetime,
     Type,
-    Const,
+    // unordered is only true if sess.has_features().const_generics
+    // is active. Specifically, if it's only `min_const_generics`, it will still require ordered
+    // types.
+    Const { unordered: bool },
+}
+
+impl PartialOrd for ParamKindOrd {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        use ParamKindOrd::*;
+        let ord = match self {
+            &Lifetime if other == &Lifetime => Ordering::Equal,
+            &Lifetime => Ordering::Less,
+
+            &Type if other == &Lifetime => Ordering::Greater,
+            &Type if other == &Type || other == &Const { unordered: true } => Ordering::Equal,
+            &Type => Ordering::Less,
+
+            &Const { .. } if other == &Lifetime => Ordering::Greater,
+
+            &Const { unordered: false } if other == &Type => Ordering::Greater,
+
+            &Const { .. } => Ordering::Equal,
+        };
+        Some(ord)
+    }
+}
+impl Ord for ParamKindOrd {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.partial_cmp(other).unwrap()
```

Comment:
wouldn't it make more sense to implement `Ord` and use `Some(self.cmp(other))` in `PartialOrd`?

Reply:
ah yea I can do that